docs/source/data/fetching-data.mdx

```diff
@@ -1,6 +1,224 @@
+
+## Creating data source classes
+
+Your data source class can be as straightforward or complex as you need it to be. You know what data your server needs, and you can let that be the guide for the methods your class should include. 
+
+Below is an example of a data source class that connects to a database:
+
+<MultiCodeBlock>
+
+```ts title="bookings.ts"
+export class BookingsDataSource {
+  private dbConnection;
+  private token;
+
+  constructor(options: { contextValue: { token: string } }) {
+    this.dbConnection = this.initializeDBConnection();
+    this.token = options.contextValue.token;
+  }
+
+  async initializeDBConnection() {
+    // set up our database details, instantiate our connection,
+    // and return that database connection
+    return dbConnection;
+  }
+
+  async getBooking(bookingId) {
+    const user = await getUser(this.token);
+    if (user) {
+      return await this.dbConnection.Booking.findByPk(bookingId);
+    } else {
+      // handle invalid user
+    }
+  }
+
+  //... more methods for finding and creating bookings
+}
+```
+
+</MultiCodeBlock>
+
+> Apollo's [`RestDataSource`](https://github.com/apollographql/datasource-rest/blob/main/src/RESTDataSource.ts) class is a built-out example of how a data source class can handle caching, deduplication, and errors.
+
+## Adding data sources to your context
+
+You can add data sources to your server's `context` initialization function, like so:
+
+<MultiCodeBlock>
+
+```ts title="index.ts"
+//highlight-start
+interface ContextValue {
+  dataSources: {
+    dogsDB: DogsDataSource;
+    catsApi: CatsAPI;
+  };
+  token: string
+}
+//highlight-end
+
+const server = new ApolloServer<ContextValue>({
+  typeDefs,
+  resolvers,
+});
+
+const { url } = await startStandaloneServer(server, {
+  context: async () => {
+    const { cache } = server; // highlight-line
+    const token = req.headers.token;
+    return {
+      // We create new instances of our data sources with each request.
+      // We can pass in our server's cache, contextValue, or any other
+      // info our data sources require. 
+      // highlight-start
+      dataSources: {
+        dogsDB: new DogsDataSource({ cache, contextValue: token }),
+        catsApi: new CatsAPI({ cache }),
+      },
+      //highlight-end
+      token
+    };
+  },
+});
+
+console.log(`🚀  Server ready at ${url}`);
+```
+
+</MultiCodeBlock>
+
+<!-- TODO(AS4) add link to context article once exists  -->
+Apollo Server calls [the `context` initialization](./resolvers/#the-context-argument) function for _every incoming operation_. This means:
+- For every operation, `context` returns an _object_ containing new instances of your data source classes (in this case, `DogsDataSource` and `CatsAPI`).
+- The **`context` function should create a new instance of each data source class for each operation.** You might accidentally combine results from multiple operations if multiple operations share a single data source instance.
```

Comment:
This is only the case if you either pass something specific to the request to the data source construction or if the data source has caching, of course... (Which is why making data source setup explicit is helpful.) Not sure if that's helpful to clarify or not. (To be honest I find the entire article could be summarized by the single sentence "fetch data using the same mechanisms you'd use to fetch data in any other server; if the code you use to do that depends on properties of the request like authentication, you can put stuff on the context value" but I do think people are probably still helped by this article.)

Reply:
Rephrased this as:

> If your data source class uses **caching** or its constructor requires an argument, the `context` function should create a _new instance_ of each data source class for each operation. Otherwise, you might accidentally combine results from multiple operations if those operations share a single data source instance.

Reply:
well I guess the issue is it's not caching in general, but "in memory caching of values that shouldn't be shared across requests".  (This is why it's a bit challenging that the REST DataSource page doesn't actually explain its two-layer caching model... you could link to that here if we had it.)

Reply:
I'll add a TODO here to add a link to the `RESTDataSource` page once it covers the two-layer caching model. For now though, how does the following feel to cover this point:
> - If your data source uses an in-memory cache, the `context` function should create a **new instance of your data source class for each operation**. This ensures that your data source doesn't accidentally share values across requests.

Reply:
Took another stab at tweaking this now that I have a better understanding of why (and I'll keep the TODO for adding the link to `RESTDataSource` once that PR is merged):
> If your data source is stateful (e.g., uses an in-memory cache), the `context` function should create a **new** instance of your data source class for **each** operation. This ensures that your data source doesn't accidentally share cached results across requests.